Add tooltip annotation functionality

kb- opened the pull request:
This PR addresses https://github.com/plotly/plotly.js/issues/7054

The tooltip function can be activated by a new modebar button.  
It allows to add an annotation to every clicked point. By default, tooltips contain x, y and when available z, or ohlc data (open, high, low, close).

Tooltips can be customized with an optional tooltiptemplate (leverages existing hovertemplate code) and tooltip annotation options (leverages existing annotations code)  
![331887266-f7258b47-6eb2-4c3c-a3ce-f23899fe57e1](https://github.com/user-attachments/assets/6ded5b0a-f01f-488f-9d38-535c2ca625bb)  
I added examples for various plot types in \test\image\mocks

### Additional interactivity:
When a plot is created with `editable: true`, the tooltips can be dragged around for placement or deleted interactively. Their text can also be edited. To delete a tooltip, click on its text and delete it.

### Compatibility: 
(active for plots with hasCartesian or hasGL2D)

- ✓ bar
- x barpolar (not active)
- **~ box (missing information in data.points[0], the tooltip is not very useful)**  
  ![image](https://github.com/user-attachments/assets/3fadc2fc-bd09-4dad-a77a-20b420aa6e0d)
- ✓ candlestick
- x carpet (not active, same for hover and spike)
- x choropleth (not active)
- x choroplethmap (not active)
- x choroplethmapbox (not active)
- x cone (not active)
- ✓ contour
- x contourcarpet (not active, same for hover and spike) 
- x densitymap (not active)
- x densitymapbox (not active)
- **~ funnel (wrong placement) Should disable? not very useful for this graph**
- x funnelarea (not active)
- ✓ heatmap
- ~~✓ heatmapgl~~(trace type no longer exists)
- ✓ histogram
- ✓ histogram2d
- ✓ histogram2dcontour
- x icicle (not active)
- x image (not active)
- x indicator (not active)
- x isosurface (not active)
- x mesh3d (not active)
- ✓ ohlc
- x parcats (not active)
- x parcoords (not active)
- x pie (not active)
- ~~✓ pointcloud~~(trace type no longer exists)
- x sankey (not active)
- ✓ scatter
- x scatter3d (not active)
- **x scattercarpet:**  
    - wrong placement, a and b missing, wrong y.  
    - note: implemented hover has a wrong y too (more obvious in scattercarpet-text)
- x scattergeo (not active)
- ✓ scattergl
- x scattermap (not active)
- x scattermapbox (not active)
- x scatterpolar (not active)
- x scatterpolargl (not active)
- x scattersmith (not active)
- x scatterternary (not active)
- ✓ splom
- x streamtube (not active)
- x sunburst (not active)
- x surface (not active)
- x table (not active, not useful)
- x treemap (not useful)
- **x violin  (missing information in data.points[0])**
- x volume (not active)
- ✓ waterfall

Usage example:
```Javascript
// Generate date and time values
function generateDateTime(start, count) {
    var dates = [];
    var current = new Date(start);
    for (var i = 0; i < count; i++) {
        dates.push(new Date(current));
        current.setHours(current.getHours() + 7);
    }
    return dates;
}

// Generate random y-values
function generateRandomYValues(count) {
    return Array.from({ length: count }, function() {
        return Math.random() * 20;
    });
}

// Generate data points
var xValues = generateDateTime('2024-04-01T12:00:00Z', 10);

var trace1 = {
    name: 'Trace 1',
    x: generateDateTime('2024-04-01T12:00:00Z', 5),
    y: generateRandomYValues(5),
    type: 'scatter',
    // Tooltip Template with: 
    // trace name, 
    // hour-minute-second, 
    // y value in exponential notation with 2 digits
    tooltiptemplate: "%{fullData.name}<br>%{x|%H:%M:%S}<br>y: %{y:.2e}",
    
    // Tooltip annotation custom format:
    tooltip: {
        align: 'left',
        arrowcolor: 'blue',
        arrowhead: 2, // Arrow type
        arrowsize: 1.2,
        arrowwidth: 1,
        font: {
            color: 'red',
            family: 'Arial',
            size: 16
        },
        showarrow: true,
        xanchor: 'left'
    }
};

var trace2 = {
    name: 'Trace 2',
    x: generateDateTime('2024-04-01T12:00:00Z', 5),
    y: generateRandomYValues(5),
    type: 'scatter',
    // Tooltip uses default template and annotation format
};

var data = [trace1, trace2];

var layout = {
    title: 'Custom Tooltip Example'
};

var config = {
    editable: true, // editable allows Tooltip drag-positioning and deletion
    modeBarButtonsToAdd: ['tooltip'] // Add the tooltip button
};

Plotly.newPlot('plot', data, layout, config);

```


# Tooltip Annotation - documentation

## Overview

Tooltip annotations let users click data points and create persistent Plotly annotations from the modebar. The feature supports three customization layers:

- `tooltiptemplate` for formatting the annotation text
- `tooltip` for annotation styling
- `tooltipfunction(ctx)` for runtime customization, remapping, or cancellation

This feature is designed for Plotly.js usage in JavaScript. The callback form is runtime-only and is attached directly to the plotted trace object.

## Requirements And Activation

To use tooltip annotations:

1. Add the tooltip modebar button with `config.modeBarButtonsToAdd: ['tooltip']`
2. Prefer `editable: true` so created annotations can be edited directly on the graph
3. Click the tooltip modebar button in the UI to enable "add on click"
4. Click a point or bin on the plot to create the annotation

Example:

```js
Plotly.newPlot('graph', data, layout, {
  editable: true,
  modeBarButtonsToAdd: ['tooltip']
});
```

`editable: true` is recommended because it makes tooltip annotations easier to work with after they are created:

- drag the annotation to a new position
- edit the text inline
- clear the text to remove the annotation

## Basic Usage With `tooltiptemplate`

`tooltiptemplate` uses the same interpolation style as `hovertemplate`. It controls the text shown inside the created annotation.

See Plotly's hovertemplate documentation:
https://plotly.com/javascript/hover-text-and-formatting/

Numeric formatting inside `%{...}` uses D3 format syntax, for example `%{x:.2f}` or `%{z:.3e}`.
See the D3 format syntax reference:
https://d3js.org/d3-format

If no `tooltiptemplate` is supplied, Plotly falls back to a built-in default based on the clicked point data, typically `x`, `y`, and `z` when available.

Basic scatter example:

```js
const data = [{
  type: 'scatter',
  mode: 'markers',
  x: [1, 2, 3],
  y: [2, 1, 4],
  tooltiptemplate: 'x: %{x:.2f}<br>y: %{y:.3f}'
}];

Plotly.newPlot('graph', data, {}, {
  editable: true,
  modeBarButtonsToAdd: ['tooltip']
});
```

Heatmap or histogram2d example:

```js
const data = [{
  type: 'heatmap',
  z: [
    [1, 2, 3],
    [2, 5, 1],
    [0, 1, 4]
  ],
  tooltiptemplate: 'x: %{x:.1f}<br>y: %{y:.1f}<br>z: %{z:.3f}'
}];
```

## Styling With `tooltip`

`tooltip` accepts annotation-style properties. It is merged with Plotly's built-in defaults for tooltip annotations.

These are the same annotation-style properties documented in Plotly's text and annotations page:
https://plotly.com/javascript/text-and-annotations/

Common styling fields include:

- `bgcolor`
- `bordercolor`
- `font`
- `arrowcolor`

Example:

```js
const data = [{
  type: 'scatter',
  mode: 'markers',
  x: [1, 2, 3],
  y: [2, 1, 4],
  tooltiptemplate: 'Point<br>x: %{x:.2f}<br>y: %{y:.2f}',
  tooltip: {
    bgcolor: 'rgba(15, 23, 42, 0.92)',
    bordercolor: '#22c55e',
    arrowcolor: '#22c55e',
    font: {
      color: '#f8fafc',
      family: 'Segoe UI, Arial, sans-serif',
      size: 12
    }
  }
}];
```

## Runtime Customization With `tooltipfunction(ctx)`

`tooltipfunction(ctx)` is the executable customization layer. It is:

- JavaScript-only
- runtime-only
- attached directly to `gd.data[traceIndex]`
- not part of serialized figure JSON

Attach it after the plot exists:

```js
const gd = document.getElementById('graph');

gd.data[0].tooltipfunction = function(ctx) {
  return {
    point: {
      extraValue: 42
    }
  };
};
```

### Return Contract

The callback may return:

- `false` or `null` to cancel tooltip creation
- a `string` to replace the template text input
- an `object` with any of these fields:
  - `point`
  - `text`
  - `annotation`
  - `style`

Meaning of each object field:

- `point`: merged into the clicked point data before `tooltiptemplate` interpolation
- `text`: replaces the resolved template string before interpolation
- `annotation`: overrides fields on the annotation object that Plotly is about to create
- `style`: adds or overrides annotation styling after the trace-level `tooltip` style is read

Examples for each field:

Example using `point`:

```js
gd.data[0].tooltiptemplate = 'x: %{x}<br>y: %{y}<br>score: %{score:.2f}';

gd.data[0].tooltipfunction = function(ctx) {
  return {
    point: {
      score: ctx.point.y * 10
    }
  };
};
```

Example returning a string:

```js
gd.data[0].tooltipfunction = function(ctx) {
  return 'Clicked x=' + ctx.point.x.toFixed(2) + '<br>Clicked y=' + ctx.point.y.toFixed(2);
};
```

Example using `text`:

```js
gd.data[0].tooltipfunction = function(ctx) {
  return {
    text: 'Formatted value<br>y = ' + ctx.point.y.toExponential(3)
  };
};
```

Example cancelling the tooltip:

```js
gd.data[0].tooltipfunction = function(ctx) {
  if(ctx.point.y < 0) return false;
  return null; // also cancels
};
```

Example injecting extra fields used by `tooltiptemplate`:

```js
gd.data[0].tooltiptemplate = 'x: %{x}<br>y: %{y}<br>note: %{note}';

gd.data[0].tooltipfunction = function(ctx) {
  return {
    point: {
      note: 'nearest sample'
    }
  };
};
```

Example using `annotation`:

```js
gd.data[0].tooltipfunction = function(ctx) {
  return {
    annotation: {
      ax: 30,
      ay: -40,
      xanchor: 'left'
    }
  };
};
```

Use `annotation` when you want to control the annotation object itself, for example:

- force a different arrow offset with `ax` and `ay`
- change the box anchor with `xanchor` or `yanchor`
- override the annotation position independently from the formatted point fields

Example using `style`:

```js
gd.data[0].tooltipfunction = function(ctx) {
  return {
    style: {
      arrowcolor: 'crimson',
      bordercolor: 'crimson'
    }
  };
};
```

## `ctx` Reference

The callback receives a single `ctx` object with these fields:

- `gd`
- `eventData`
- `event`
- `point`
- `trace`
- `fullTrace`
- `calcdata`
- `fullLayout`
- `xaxis`
- `yaxis`

Practical usage of each field:

### `ctx.gd`

The graph div. Use this when you need direct access to the rendered plot object.

```js
console.log(ctx.gd.data.length);
```

### `ctx.eventData`

The `plotly_click` payload. Useful when you want the full click event object, including all clicked points.

```js
console.log(ctx.eventData.points);
```

### `ctx.event`

The raw mouse event. Useful for screen-space calculations such as `clientX`, `clientY`, or DOM target geometry.

```js
const rect = ctx.event.target.getBoundingClientRect();
console.log(ctx.event.clientX - rect.left, ctx.event.clientY - rect.top);
```

### `ctx.point`

The clicked point or bin event data. This is usually the main input for runtime tooltip logic.

Typical values include:

- `x`
- `y`
- `z`
- `pointIndex` or `pointIndices`
- `pointNumbers`
- `customdata` when available

```js
console.log(ctx.point.x, ctx.point.y, ctx.point.z);
```

### `ctx.trace`

The original input trace from `gd.data`. This is where runtime-only additions such as `tooltipfunction` live.

```js
console.log(ctx.trace.name);
console.log(ctx.trace.customdata);
```

### `ctx.fullTrace`

The fully processed trace from `gd._fullData`. It is the plotted version of the trace after Plotly has applied defaults and internal preprocessing. Use it when you need the trace as Plotly is actually using it, not only the raw input you provided in `gd.data`.

```js
console.log(ctx.fullTrace.tooltiptemplate);
console.log(ctx.fullTrace.type);
```

### `ctx.calcdata`

Trace-specific calcdata. Useful for advanced logic on binned traces such as `histogram2d`, where you may want access to computed bins and values.

```js
const cd0 = ctx.calcdata[0];
console.log(cd0.z);
```

### `ctx.fullLayout`

The fully processed layout object from the rendered plot. Use it when you need the active layout state that Plotly is currently using, such as the current locale, axis objects, hover mode, subplot internals, or computed defaults.

```js
console.log(ctx.fullLayout.hovermode);
console.log(ctx.fullLayout._d3locale);
```

### `ctx.xaxis` and `ctx.yaxis`

The clicked subplot axes. These are Plotly axis objects with conversion helpers.

Common internal helpers include:

- `c2p`: data coordinate to pixel
- `p2c`: pixel to data coordinate
- `d2c`: displayed value to coordinate, often useful on categorical axes

These helpers are internal Plotly axis methods rather than public top-level API methods, so they are best treated as advanced usage helpers inferred from the current implementation.

```js
const xPixel = ctx.xaxis.c2p(ctx.point.x);
const yPixel = ctx.yaxis.c2p(ctx.point.y);
console.log(xPixel, yPixel);
```

See Plotly's event data documentation for the general click-event payload shape:
https://plotly.com/javascript/plotlyjs-events/

## `customdata` In The Callback

For direct point-based traces, `customdata` is often available directly on the clicked point:

```js
gd.data[0].tooltipfunction = function(ctx) {
  console.log(ctx.point.customdata);
  return {
    point: {
      label: ctx.point.customdata
    }
  };
};
```

For aggregated traces like `histogram2d`, a clicked bin usually corresponds to many source samples. In that case, use `ctx.point.pointIndices` with `ctx.trace.customdata`:

```js
gd.data[0].tooltipfunction = function(ctx) {
  const indices = ctx.point.pointIndices || [];
  const rawCustom = indices.map(function(i) {
    return ctx.trace.customdata[i];
  });

  return {
    point: {
      sampleCount: indices.length,
      firstCustomValue: rawCustom[0]
    }
  };
};
```

## Examples

### Example 1: Basic Scatter Tooltip Annotation

```js
const trace = {
  type: 'scatter',
  mode: 'markers',
  x: [1, 2, 3],
  y: [2, 1, 4],
  tooltiptemplate: 'x: %{x:.2f}<br>y: %{y:.2f}',
  tooltip: {
    arrowcolor: 'blue'
  }
};

Plotly.newPlot('graph', [trace], {}, {
  editable: true,
  modeBarButtonsToAdd: ['tooltip']
});
```

### Example 2: Programmatic Attachment On An Existing Plot

```js
const gd = document.getElementById('graph');

gd.data[0].tooltiptemplate =
  'x: %{x:.2f}<br>y: %{y:.2f}<br>score: %{score:.1f}';

gd.data[0].tooltipfunction = function(ctx) {
  return {
    point: {
      score: ctx.point.y * 10
    }
  };
};
```

### Example 3: Styling A Tooltip Annotation

```js
const gd = document.getElementById('graph');

gd.data[0].tooltip = {
  bgcolor: 'rgba(255, 255, 255, 0.95)',
  bordercolor: '#2563eb',
  arrowcolor: '#2563eb',
  font: {
    color: '#111827',
    family: 'Georgia, serif',
    size: 13
  }
};
```

### Example 4: Histogram2d Or Heatmap Local Maximum

This pattern remaps the tooltip to the strongest value inside a rectangular kernel.

In this example:

- `kernelSizeX` and `kernelSizeY` are in plot data units, not pixels
- the callback searches nearby heatmap or histogram2d cells
- returning new `point.x`, `point.y`, and `point.z` changes both the formatted values and the default arrow anchor position

If you only return `point.x` and `point.y`, the arrow already moves to that remapped point. You only need `annotation` as well if you want the annotation object itself to differ, for example with a custom `ax`, `ay`, `xanchor`, or a deliberately different `x` / `y` than the remapped point.

```js
const gd = document.getElementById('graph');

gd.data[0].tooltiptemplate =
  'Local max: %{z:.4f}<br>x: %{x:.3f}<br>y: %{y:.3f}<br>kernel: %{kernelSizeX} x %{kernelSizeY}';

gd.data[0].tooltipfunction = function(ctx) {
  // Kernel size in plot units, not pixels.
  const kernelSizeX = 3;
  const kernelSizeY = 3;

  // For heatmap-like traces, Plotly stores the plotted z matrix in calcdata.
  // If _x/_y are present on the full trace, they are the plotted x/y centers.
  const z = ctx.fullTrace._z || ctx.calcdata[0].z;
  const xs = (ctx.fullTrace._x && ctx.fullTrace._x.length) ?
    ctx.fullTrace._x :
    ctx.calcdata[0].xRanges.map(r => (r[0] + r[1]) / 2);
  const ys = (ctx.fullTrace._y && ctx.fullTrace._y.length) ?
    ctx.fullTrace._y :
    ctx.calcdata[0].yRanges.map(r => (r[0] + r[1]) / 2);

  // Search bounds centered on the clicked point.
  const minX = ctx.point.x - kernelSizeX / 2;
  const maxX = ctx.point.x + kernelSizeX / 2;
  const minY = ctx.point.y - kernelSizeY / 2;
  const maxY = ctx.point.y + kernelSizeY / 2;

  let best = -Infinity;
  let bestX = ctx.point.x;
  let bestY = ctx.point.y;

  // Scan the bins whose centers fall inside the kernel rectangle.
  for(let iy = 0; iy < ys.length; iy++) {
    if(ys[iy] < minY || ys[iy] > maxY) continue;
    for(let ix = 0; ix < xs.length; ix++) {
      if(xs[ix] < minX || xs[ix] > maxX) continue;
      const value = z[iy][ix];
      if(value > best) {
        best = value;
        bestX = xs[ix];
        bestY = ys[iy];
      }
    }
  }

  return {
    // These remapped point fields drive both template interpolation and
    // the default arrow anchor position.
    point: {
      x: bestX,
      y: bestY,
      z: best,
      kernelSizeX,
      kernelSizeY
    }
  };
};
```

If you want the annotation object itself to differ from the remapped point, add `annotation` explicitly:

```js
return {
  point: {
    x: bestX,
    y: bestY,
    z: best,
    kernelSizeX,
    kernelSizeY
  },
  annotation: {
    ax: 40,
    ay: -30
  }
};
```

### Example 5: Cancellation

Returning `false` or `null` is useful for conditional display when only some points or bins should produce tooltip annotations.

```js
gd.data[0].tooltipfunction = function(ctx) {
  if(ctx.point.z !== undefined && ctx.point.z < 0) {
    return false;
  }

  return {
    point: {
      accepted: true
    }
  };
};
```

## Caveats And Notes

- `tooltipfunction(ctx)` is synchronous only
- the callback is runtime-only and not JSON-serializable
- wrapper languages such as Python or R need custom injected JavaScript if they want callback behavior
- tooltip annotations are real layout annotations, so adding them may trigger noticeable redraw latency on very large heatmaps or other heavy figures
- the modebar tooltip button must be enabled and activated by the user before click-to-add behavior starts

kb- committed:
Init - basic working implementation

Implemented style and template handling

kb- committed:
Implemented style and template handling

kb- committed:
Added Dash and Python demo

kb- committed:
Update index.html

kb- committed:
Update README.md

kb- committed:
Update README.md

kb- committed:
Added schema and cleanup

kb- committed:
Made linter happy

kb- committed:
Update README.md

kb- committed:
locales

kb- committed:
Merge branch 'dev-tooltip' of plotly.js:kb-/plotly.js into dev-tooltip

kb- committed:
Demo

kb- committed:
Update various.html

kb- committed:
Update various.html

kb- committed:
Update various.html

kb- committed:
Handle log axis

kb- committed:
Handle log axis

kb- committed:
Check parameters availability

kb- committed:
Merge branch 'dev-tooltip' of plotly.js:kb-/plotly.js into dev-tooltip

kb- committed:
Merged src/components/modebar/buttons.js from dev-tooltip-all into dev-tooltip

kb- committed:
Merge branch 'master'

kb- committed:
build

kb- committed:
Added test mocks

kb- committed:
untrack dist

kb- committed:
Added Jasmine test

kb- committed:
linter fix

kb- committed:
Merge branch 'master' into dev-tooltip

kb- committed:
cleanup

kb- committed:
Add histogram2d test mock

kb- committed:
fixed 0 value handling

kb- committed:
fixed multi curve histogram handling

kb- committed:
Handle subplots

kb- committed:
make default number format consistent with hover

kb- committed:
Add z to heatmap tooltip default template

kb- committed:
Fixed subplot handling

kb- committed:
Fixed test

kb- committed:
generate histogram test click

kb- committed:
Change to assets click function

kb- committed:
Added test on log axis plot

kb- committed:
fixed timing

kb- committed:
Test preventing second tooltip on same data point

kb- committed:
Generic default template based on data availability

kb- committed:
Complete attributes

kb- committed:
Handle multycategory

kb- committed:
Merge branch 'master' into dev-tooltip

kb- committed:
fix subplot handling

kb- committed:
Added default template for category axis

Revert "Added default template for category axis"

This reverts commit 99ed6e7c10b996b55bff4630fd955269cba26085.

Reapply "Added default template for category axis"

This reverts commit 0572392854029b1e79701e46a8003da99d97a282.

Revert "Added default template for category axis"

This reverts commit 99ed6e7c10b996b55bff4630fd955269cba26085.

Added default template for category axis

kb- committed:
More test image mocks

kb- committed:
Add violin side to side handling

kb- committed:
Merge branch 'master' into dev-tooltip

kb- committed:
Fixed test click point

kb- committed:
handle missing axis info

kb- committed:
Added tooltip to hasGL2D

kb- committed:
Added gl2d mocks

kb- committed:
Schema update

kb- committed:
Blacklist mock files derived from existing blacklisted files

kb- committed:
test_syntax

kb- committed:
Merge branch 'master' into dev-tooltip

kb- committed:
fixed attributes - plotschema pass

kb- committed:
removed circular reference

kb- committed:
add baselines (no tooltip displayed)

kb- committed:
draftlog

kb- committed:
test-mock pass

kb- committed:
Implement heatmap attribute - default tooltiptemplate

kb- committed:
schema

kb- commented:
This is my first contribution to a large project like this. 

Most CI tests pass, but I need some help for the mock-validation test. _test\image\mocks\tooltip_heatmap.json_
It looks like I have to add the tooltip and tooltiptemplate attributes to each trace type attributes.js. 

I've been able to add tooltiptemplate (string) to the heatmap trace attributes. _src\traces\heatmap\attributes.js_
But I don't know how to add tooltip (object; it should accept the same attributes than annotations)

> node tasks/test_mock.js tooltip_heatmap
> 
> validating tooltip_heatmap
> Expected false to be true
> file: tooltip_heatmap
> [
>   {
>     "code": "unused",
>     "container": "data",
>     "trace": 0,
>     "path": [
>       "tooltip"
>     ],
>     "astr": "tooltip",
>     "msg": "In data trace 0, container tooltip did not get coerced"
>   }
> ]
> 
> C:\Users\olivi\Documents\code\plotly.js\tasks\test_mock.js:44
>     throw error;
>     ^
> Failed at {
>   "mocks": [
>     "tooltip_heatmap"
>   ]
> }
> (Use `node --trace-uncaught ...` to show where the exception was thrown)

gvwilson commented:
@kb- I'll try to get someone to have a look next week - thanks.

kb- committed:
Create tooltip_template_heatmap.png

kb- committed:
Merge branch 'master' into dev-tooltip

kb- committed:
fixed heatmat attributes

kb- committed:
fixed scatter attributes

kb- committed:
fixed bar attributes

kb- committed:
fixed bar attributes

kb- committed:
fixed box defaults

kb- committed:
fixed funnel defaults

kb- committed:
fixed candlestick attributes

kb- committed:
fixed heatmapgl defaults

kb- committed:
fixed scattergl defaults

kb- committed:
fixed histogram attributes

kb- committed:
fixed histogram2d attributes

kb- committed:
fixed histogram2dcontour attributes

kb- committed:
fixed ohlc attributes (bug on master)

kb- committed:
fixed pointcloud defaults (bug on master)

kb- committed:
fixed scattercarpet attributes

kb- committed:
fixed contour defaults

kb- committed:
fixed splom attributes

kb- committed:
fixed violin defaults

kb- committed:
fixed waterfall attributes

kb- committed:
fixed attributes

kb- committed:
Merge branch 'master' into dev-tooltip

kb- committed:
schema

kb- committed:
removed ohlc tests (ohlc_first from master fails too)

kb- committed:
baselines

kb- committed:
removed obsolete plot types tests

kb- committed:
removed candlestick test (candlestick from master fails too)

kb- committed:
removed obsolete plot types test

kb- commented:
> @emilykl Can you help me fix the remaining issues?

I figured it out.
These tests don't pass on master ` node tasks/test_mock.mjs candlestick_double-y-axis ohlc_first`, so I removed the tests I had based on them (tooltip_tooltiptemplate_candlestick_double-y-axis ce08efbbd66a2a0a0ea8e561d20828b62a03df17 and tooltip_tooltiptemplate_ohlc baf913ed17ff0f389c28f1641a94c7f99101b897)

kb- committed:
removed dflt (was not compatible with category)

kb- committed:
fixed scattercarpet attributes

kb- committed:
fixed heatmap dflt (was not compatible with category)

kb- committed:
fixed bar arrow placement for shared categories

kb- committed:
schema

kb- committed:
fixed jasmine test

kb- committed:
baseline

kb- committed:
moved ohlc template to attributes

kb- committed:
fixed candlestick attributes

kb- committed:
mocks

kb- committed:
fixed waterfall arrow placement

kb- committed:
schema

kb- committed:
Delete notes.txt

kb- committed:
Resolved merge conflict

kb- committed:
schema

kb- committed:
Handle stacked traces

kb- committed:
baselines

kb- committed:
removed mock

gvwilson commented:
Apologies @kb- - we are trying to get the next Dash and Plotly.py releases out the door, so we're short on cycles for reviewing PRs (particularly ones that touch this many files). I'll try to prioritize it in the next work cycle, but realistically it's going to be at least a couple more weeks :-( Thanks for your patience - @gvwilson

kb- committed:
Merge remote-tracking branch 'upstream/master' into dev-tooltip-conflict-fix

kb- committed:
Fixed tooltip annotation text rendering and regenerated schema

kb- commented:
Rebased this PR onto current upstream and fixed the tooltip annotation modebar so it works again after the template-formatting changes that landed since this was opened.

What changed:
- resolved the merge conflicts against current upstream
- updated tooltip annotation text rendering to use the current `Lib.hovertemplateString` API
- treated empty `tooltiptemplate` values as unset so the generated default tooltip text still appears
- regenerated `test/plot-schema.json`

Verification:
- checked tooltip annotation behavior in the test dashboard
- ran `npm run schema`
- confirmed `tooltip_annotation_test.js` passes

This should now be back in a reviewable state. I realize priorities shift, but since the PR has been dormant for a year, I wanted to surface it again and hope you still consider it. Thanks!

kb- committed:
Added tooltip callback function

kb- commented:
I added an optional callback function. The tooltip feature is now more powerful than the similar one in Matlab😎
[tooltipfunction usage demo](https://kb-.github.io/Plotly.js_tooltip/)
